blockchair/src/endpoint/balance.ts

```diff
@@ -0,0 +1,147 @@
+import objectPath from 'object-path'
+import { Requester, Validator } from '@chainlink/external-adapter'
+import { AdapterRequest } from '@chainlink/types'
+import { Config, DEFAULT_DATA_PATH, getBaseURL } from '../config'
+import { CoinType, ChainType } from '.'
+
+export const Name = 'balance'
+
+type Address = {
+  address: string
+  coin?: CoinType
+  chain?: ChainType
+  balance?: number
+  warning?: string
+}
+
+type AddressGroup = {
+  addresses: string[]
+  result: Address[]
+}
+
+type GroupedAddresses = {
+  [network: string]: AddressGroup
+}
+
+type ResponseWithResult = {
+  response: any
+  result: Address[]
+}
+
+type RequestData = {
+  dataPath: string
+  confirmations: number
+}
+
+const coins: { [ticker: string]: string } = {
+  btc: 'bitcoin',
+  dash: 'dash',
+  doge: 'dogecoin',
+  ltc: 'litecoin',
+  zec: 'zcash',
+  bch: 'bitcoin-cash',
+  bsv: 'bitcoin-sv',
+  grs: 'groestlcoin',
+}
+
+const WARNING_NO_OPERATION_TESTNET =
+  'No Operation: only testnet supported by blockcypher adapter is btc'
+const WARNING_NO_OPERATION_MISSING_ADDRESS = 'No Operation: address param is missing'
+const WARNING_NO_OPERATION_ADDRESS_NOT_FOUND = 'No Operation: address does not exist on network'
+
+const group = (addresses: Address[]) => {
+  const output: GroupedAddresses = {}
+  for (const addr of addresses) {
+    if (!addr?.address) addr.warning = WARNING_NO_OPERATION_MISSING_ADDRESS
+    if (!addr?.coin) {
+      addr.coin = 'btc'
+    }
+    if (!addr?.chain) {
+      addr.chain = 'mainnet'
+    }
+    const { address, coin, chain } = addr
+    if (output[`${coin}-${chain}`]) {
+      output[`${coin}-${chain}`].addresses.push(address)
+      output[`${coin}-${chain}`].result.push(addr)
+    } else output[`${coin}-${chain}`] = { addresses: [address], result: [addr] }
+  }
+  return output
+}
+
+const getBalanceURI = (jobRunID: string, addresses: string[], chain: string, coin: string) => {
+  coin = coins[coin]
+  if (!coin) throw Requester.errored(jobRunID, `Unsupported coin parameter`, 400)
+  if (chain === 'testnet') {
+    if (coin === 'bitcoin') coin = 'bitcoin/testnet'
+    else throw Requester.errored(jobRunID, WARNING_NO_OPERATION_TESTNET, 400)
+  }
+  return `/${coin}/addresses/balances?addresses=${addresses.join(',')}`
+}
+
+const getBalances = async (
+  jobRunID: string,
+  config: Config,
+  groupedData: GroupedAddresses,
+): Promise<ResponseWithResult[]> =>
+  Promise.all(
+    Object.entries(groupedData).map(async ([network, reqData]) => {
```

Comment:
It looks like we are using Blockchairs batch endpoint, so `getBalanceURI(jobRunID, reqData.addresses, chain, coin)` will return an endpoint that will fetch us balances for all addresses. Why do we do it N times then?

Reply:
Aha, do we group addresses for different chains for example `bitcoin/testnet` & `bitcoin/mainet` and then we execute M requests, where M is the number of different chains?

Reply:
Yep exactly. Blockchair has an [endpoint](https://blockchair.com/api/docs#link_390) for getting multiple addresses from a certain chain. In this case it is **extraordinarily** cheaper to use.